### ISSUE TITLE
ArgumentConstraintTrap is not threadsafe

### ISSUE THREAD

**Issue by blairconrad** [FakeItEasy/FakeItEasy#476]
This is one of the problems preventing us from supporting parallel test execution out of the box (#60), but may not be the only factor, so I preferred to start a new issue. It is the only concrete 

The internal `ArgumentConstraintTrap` class, used to capture argument constraints using the `That` syntax (`A<SomeClass>.That.Matches(…)`), is not threadsafe. It uses a single static variable to keep trap of the captured arguments. This can cause parallel test execution to fail, whether the `That` is used to configure a fake or to assert that certain calls were made to the fake.

I believe that replacing the static `List<IArgumentConstraint>` with a `ThreadLocal<List<IArgumentConstraint>>` will fix the problem, and preliminary testing has supported this.
Of course, testing never _proves_ anything, but I think making the field ThreadLocal is a reasonable step.

**Comment by blairconrad** [comment]
The unsafety of the `ArgumentConstraintTrap` appears to affect more configuration/assertions than I thought, even this kind of config:

```c#
A.CallTo(() => sut.SomethingMethod()).Returns(true);
```

Fortunately, the one fix seems to be doing the trick for that too.

**Comment by adamralph** [comment]
Thanks for narrowing this down!

However, I don't think `ThreadLocal` is enough what if the fake is passed out to other threads during the call to the SUT?

**Comment by blairconrad** [comment]
I believe that the constraint trap is only read from during the `That` (and by extension, `Ignore`) calls. It's written to in more cases, but only for the purpose of building up a list that can be read from. Looking at `TrapConstraints`, we set up the listener, invoke the action that produces the constraint, and then clear the `trappedConstraints`:

```c#
public IEnumerable<IArgumentConstraint> TrapConstraints(Action actionThatProducesConstraint)
{
    trappedConstraints = new List<IArgumentConstraint>();
    var result = trappedConstraints;

    actionThatProducesConstraint.Invoke();

    trappedConstraints = null;
    
    return result;
}
```

I think we're staying the same thread unless someone supplies a matcher that jumps on a different thread to perform some work, and honestly, they're asking for it. Unless I'm completely off base, all the code in the block I pasted should execute in the same thread.

You may ask how, then, that works with `A.CallTo(() => sut.SomethingMethod()).Returns(true);`. I'm not sure. Maybe the ThreadLocal just happened to make things better, by luck. I whipped up that test and ran out the door, so don't remember what kind of exception was being thrown.
Although for the test that used that kind of call, as well as for 2 tests that used `That` for config and assertion respectively, switching between the original implementation and my ThreadLocal one consistently had the desired effect: classic implementation always failed and the ThreadLocal one always passed, for the 5 or so test runs I executed (I know! What rigour!)

**Comment by blairconrad** [comment]
Remember when I said

> You may ask how, then, that works with A.CallTo(() => sut.SomethingMethod()).Returns(true);. I'm not sure.

Turns out there was a flaw in my test. Ignore that part. Everything else I said before is still true, though.

**Comment by blairconrad** [comment]
Of course, there's no `ThreadLocal` in 3.5 or SL, but there's [ThreadStatic](https://msdn.microsoft.com/en-us/library/system.threadstaticattribute(v=vs.90).aspx), which is probably as good. Better, maybe.

**Comment by blairconrad** [comment]
I'd like to fix this in 2.0. It's not a breaking change, but I like the parallel of better supporting xUnit 2.0 with FakeItEasy 2.0

**Comment by ursenzler** [comment]
I may have run into the same problem. 
I updated from xUnit 1.9 to 2.0 and I have failing tests. All these tests use a matcher on an argument (That). 
What worries me most however is, that I disabled parallelization ([assembly: CollectionBehavior(DisableTestParallelization = true)]) in xUnit, but the problem remains ?!?!

**Comment by blairconrad** [comment]
@ursenzler, I'm not familiar with xUnit option, but assume you used the option properly. I'm disturbed that the problem remains. Are you able to share anything more, such as a small example showing the problem?

**Comment by blairconrad** [comment]
Oh. After [reading about parallelism in runners](http://xunit.github.io/docs/running-tests-in-parallel.html#parallelism-in-runners), I wonder if the runner isn't still executing the tests in parallel. Sorry if I'm making no sense at all. I have nearly no experience with xUnit, having only used it once. To reproduce this issue, as it turns out.

**Comment by ursenzler** [comment]
I added the [ThreadStatic] to the trappedConstraints

This solve the issue I have.

I don't know whether xUnit 2.0 runs things on different threads even when parallelization is turned off.

**Comment by blairconrad** [comment]
@ursenzler, in the meantime, are you interested in taking an early build with what I propose as the fix for this issue? You should be able to [download the nupkg from TeamCity](http://teamcity.codebetter.com/viewLog.html?buildId=188348&buildTypeId=bt929&tab=artifacts#!132is1d).

Oh! I suspect that Scopes aren't threadsafe either, but haven't gotten around to trying them out.

**Comment by ursenzler** [comment]
I can try it if you want, but I have to wait for xBehave running on xUnit 2.0 anyways. I'm just experimenting at the moment.

**Comment by ursenzler** [comment]
Just tried it, looks good.
I don't use scopes, though.

**Comment by blairconrad** [comment]
> I added the `[ThreadStatic]` to the `trappedConstraints`

Yeah, me too.  But you probably read that.

> I can try it if you want, but I have to wait for xBehave running on xUnit 2.0 anyways. I'm just experimenting at the moment.

Oh, thanks, but don't go out of your way. It sounds like you've end up with the same code anyhow.

Thanks for your understanding and for pursuing this in xunit/xunit#392. I'm also quite curious about why disabling the parallelization had no effect.

**Comment by blairconrad** [comment]
Darn it! We are crossing wires.

Thanks for trying the package. Yeah, I'm not a scope-user, either.

**Comment by bradwilson** [comment]
FYI, thready safety is not enough for xUnit.net 2.

You need to use `CallContext` to correctly get and set values that will tracked to a given `Task`. Since tasks often will live their lives on several threads, a simple thread static will fail. xUnit.net 2 not only allows users to write async tests, it also is heavily `Task`-based internally.

The APIs to use are `LogicalGetData` and `LogicalSetData`:

https://msdn.microsoft.com/en-us/library/system.runtime.remoting.messaging.callcontext.logicalgetdata
https://msdn.microsoft.com/en-us/library/system.runtime.remoting.messaging.callcontext.logicalsetdata

They are available in .NET 2.0.

Note that since these things are designed to be remoted, everything you put into the call context must be serializable. When we hacked a fix into NSubstitute to make it work, rather than making everything serializable, we used a mapping dictionary. This isn't necessarily the best strategy, because now you have orphaned objects in the dictionary living longer lives than necessary, but it was good enough to get us running again. Ideally, the things you store there should be truly serializable, so that they go away when the call context is finished.

https://github.com/CenturyLinkCloud/NSubstitute/blob/f7eee68840eace09b757c500fadd47839222d22d/Source/NSubstitute/Core/RobustThreadLocal_Net45.cs

**Comment by blairconrad** [comment]
Thanks, @bradwilson. New concepts for me, which I'll have to absorb. I appreciate you putting the time into this.

I do wonder about the special case of the ArgumentConstraintTrap, though. The value that's being made `ThreadStatic` is extremely short-lived, and are used only in the test setup or verification phases.

Perhaps I should do my own learning, but I would not expect the code to jump threads at this time. Or am I way off-base and we may jump threads between any instructions?

**Comment by bradwilson** [comment]
You can jump threads any time there is an `await` keyword in the method. The [original bug report](https://github.com/xunit/xunit/issues/357) included this as a repro:

```csharp
var sut = A.Fake<IAsyncClass>();
await sut.GetAsync(new Input{Test = "test"});
A.CallTo(() => sut.GetAsync(A<Input>.That.Matches(_ => _.Test == "test"))).MustHaveHappened();
```

**Comment by blairconrad** [comment]
Thanks for that.

If `await` is the trigger, then I think we're safe for sane uses of `ArgumentConstraintTrap` (one could come up with a contrived example that fails, but that's another story).
I'm convinced there are other problems in FakeItEasy, but for that one use, we should be good.
The `ArgumentConstraintTrap` is only used during the 3rd line in that example, while the `A.CallTo` method is in flight. The `await` is well over by the time the code we're worried about is executed. In this case, the `await` is incidental to the problem. I would not have expected you to follow the thread over in the FakeItEasy issue tracker, but I was able to [take the repro code and de-Taskify it, reproducing without using await](https://github.com/FakeItEasy/FakeItEasy/issues/60#issuecomment-89285533).

I'm not saying that pursuing `CallContext` is a bad idea, just that for this case, it may not be necessary.

**Comment by blairconrad** [comment]
The `RobustThreadLocal` implementation is quite clever.

I see there's no `CallContext` in SilverLight. Not a dealbreaker, but it would complicate things a little.

**Comment by adamralph** [comment]
@ursenzler off topic here, but it would be great if you could try out xbehave 2.0 RC1. I'm currently in the process of upgrading several projects to it as a test before hitting the button on RTM. All feedback welcome at https://github.com/xbehave/xbehave.net/issues/144

**Comment by ursenzler** [comment]
@adamralph I just did that, I put my feedback unfortunately in issue https://github.com/xbehave/xbehave.net/issues/51

**Comment by blairconrad** [comment]
@ursenzler, thanks very much for your work on this issue. Look for your name in the release notes. :trophy:

This issue has been fixed in [release 2.0.0](https://github.com/FakeItEasy/FakeItEasy/releases/tag/2.0.0).
